File: euxis-core/src/euxis_core/mesh/swarm.py
# SPDX-License-Identifier: AGPL-3.0-or-later
# Copyright (c) 2024-2026 Sebastien Rousseau ᛫ https://sebastienrousseau.com

"""Swarm Orchestrator for Multi-Agent Playbook execution."""

from __future__ import annotations

import json
import logging
import uuid
from dataclasses import dataclass, field
from typing import Any, Dict, List, Optional

LOGGER = logging.getLogger("euxis.core.swarm")

from .router import FinOpsRouter
from euxis_core.resilience import CircuitBreaker, RetryPolicy
from euxis_core.runtime import (
    GatewayWebSocketClient,
    gather_awaitables,
    run_async_with_resilience,
)

@dataclass
class SwarmTask:
    """A single task within a swarm phase."""
    agent_id: str
    template: str
    status: str = "pending"
    result: Optional[str] = None
    run_id: str = field(default_factory=lambda: f"run_{uuid.uuid4().hex[:8]}")
    complexity: str = "medium"

class SwarmOrchestrator:
    """Orchestrates complex multi-agent workflows defined in playbooks."""

    def __init__(self, gateway_url: str, token: Optional[str] = None):
        # Convert http(s) to ws(s) for gateway communication
        self.gateway_url = gateway_url.replace("http://", "ws://").replace("https://", "wss://")
        self.token = token
        self.history: List[Dict[str, Any]] = []
        self.router = FinOpsRouter()
        self.ws_client = GatewayWebSocketClient(self.gateway_url, token=self.token)
        self.connection_retry_policy = RetryPolicy(
            max_attempts=3, base_delay_seconds=0.1, max_delay_seconds=1.0, jitter_ratio=0.0
        )
        self.connection_breaker = CircuitBreaker(failure_threshold=3, recovery_timeout_seconds=5.0)
        self.simulation_mode = "localhost" in self.gateway_url

    async def execute_playbook(self, playbook: Dict[str, Any], goal: str):
        """Execute all phases of a playbook."""
        LOGGER.info(f"Starting swarm playbook: {playbook.get('name')}")
        
        try:
            for phase in playbook.get("phases", []):
                await self._execute_phase(phase, goal)
        finally:
            await self.ws_client.close()
            
        LOGGER.info("Swarm playbook execution complete.")
        return self.history

    async def _execute_phase(self, phase: Dict[str, Any], goal: str):
        """Execute a single phase (parallel or sequential)."""
        name = phase.get("name", "Unknown Phase")
        mode = phase.get("mode", "sequential")
        delegates = phase.get("delegates", [])
        
        LOGGER.info(f"Executing phase: {name} (mode: {mode})")
        
        tasks = []
        for d in delegates:
            content = d.get("task_template", "").replace("${goal}", goal)
            # Replace history if present
            if "${history}" in content:  # pragma: no cover
                content = content.replace("${history}", json.dumps(self.history))  # pragma: no cover
            
            tasks.append(SwarmTask(d.get("agent"), content))

        if mode == "parallel":
            await gather_awaitables(self._run_task(t) for t in tasks)
        else:
            for t in tasks:
                await self._run_task(t)

    async def _ensure_connection(self):
        """Lazily establish a WebSocket connection to the gateway."""
        if self.ws_client.closed:
            try:
                await run_async_with_resilience(
                    self.ws_client.connect,
                    retry_policy=self.connection_retry_policy,
                    circuit_breaker=self.connection_breaker,
                )
            except Exception as e:
                status_code = self.ws_client.invalid_status_code(e)
                if status_code == 401:
                    LOGGER.error(f"Gateway authentication failed (401). Check your token.")
                else:
                    LOGGER.error(f"Failed to connect to gateway at {self.gateway_url}: {e}")
                raise

    async def _run_task(self, task: SwarmTask):
        """Dispatch a task to the Euxis Gateway via WebSocket and wait for result."""
        provider = self.router.select_provider(task.complexity)
        LOGGER.info(f"Dispatching task to agent {task.agent_id} via {provider}...")
        task.status = "running"

        if self.simulation_mode:
            task.status = "completed"
            task.result = f"Simulated result from {task.agent_id}"
            self.router.track_usage(provider, 500)
            self.history.append(
                {
                    "agent": task.agent_id,
                    "provider": provider,
                    "task": task.template,
                    "result": task.result,
                }
            )
            return
        
        await self._ensure_connection()
        
        req_id = task.run_id
        payload = {
            "type": "request",
            "id": req_id,
            "method": "chat.send",
            "params": {
                "session_id": f"swarm_{req_id}",
                "role": "user",
                "content": task.template,
                "meta": {
                    "agent": task.agent_id, 
                    "provider": provider,
                    "approved": True
                }
            }
        }
        
        try:
            # Send the request
            await self.ws_client.send(json.dumps(payload))
            
            # Read messages until we get a final event for this run_id
            async for message in self.ws_client.iter_messages():
                data = json.loads(message)
                
                # Check for error result
                if data.get("type") == "result" and data.get("id") == req_id:
                    if "error" in data:
                        raise Exception(f"Gateway error: {data['error']}")
                
                # Check for agent events
                if data.get("type") == "event" and data.get("event") == "agent":
                    evt_data = data.get("data", {})
                    # Only care about our run_id
                    if evt_data.get("run_id") == req_id:
                        status = evt_data.get("status")
                        if status == "final":
                            task.status = "completed"
                            task.result = evt_data.get("content")
                            # Extract token usage if available from metadata
                            tokens = evt_data.get("meta", {}).get("usage", {}).get("total_tokens", 500)
                            self.router.track_usage(provider, tokens)
                            break
                        elif status == "error":
                            raise Exception(f"Agent error: {evt_data.get('content')}")
            
            if task.status != "completed":
                raise Exception("WebSocket closed before task completion")

            self.history.append({
                "agent": task.agent_id,
                "provider": provider,
                "task": task.template,
                "result": task.result
            })
            
        except Exception as e:
            LOGGER.error(f"Task {task.run_id} failed: {e}")
            task.status = "failed"
            task.result = f"Error: {e}"
            self.history.append({
                "agent": task.agent_id,
                "provider": provider,
                "task": task.template,
                "result": task.result
            })

File: euxis-cpp/euxis-core-cpp/src/swarm.cpp
#include "euxis/core/swarm.hpp"
#include "euxis/core/ws_client.hpp"

#include <future>
#include <random>
#include <spdlog/spdlog.h>

namespace euxis::core {
namespace {

auto generate_run_id() -> std::string {
    static std::mt19937 gen(std::random_device{}());
    static std::uniform_int_distribution<uint32_t> dist;
    char buf[16];
    std::snprintf(buf, sizeof(buf), "run_%08x", dist(gen));
    return buf;
}

/// Derive the WebSocket URL from an HTTP gateway URL.
/// e.g. "http://example.com:8080" -> "ws://example.com:8081"
auto derive_ws_url(const std::string& gateway_url) -> std::string {
    // Replace http:// with ws://
    std::string ws_url = gateway_url;
    if (ws_url.starts_with("https://")) {
        ws_url = "wss://" + ws_url.substr(8);
    } else if (ws_url.starts_with("http://")) {
        ws_url = "ws://" + ws_url.substr(7);
    } else {
        ws_url = "ws://" + ws_url;
    }

    // Bump port by 1 (HTTP port -> WS port)
    auto colon = ws_url.rfind(':');
    if (colon != std::string::npos && colon > 5) {
        auto port_str = ws_url.substr(colon + 1);
        // Strip trailing path
        auto slash = port_str.find('/');
        if (slash != std::string::npos) {
            port_str = port_str.substr(0, slash);
        }
        try {
            int port = std::stoi(port_str) + 1;
            ws_url = ws_url.substr(0, colon + 1) + std::to_string(port);
        } catch (...) {
            // Leave as-is
        }
    }
    return ws_url;
}

} // namespace

SwarmOrchestrator::SwarmOrchestrator(const std::string& gateway_url,
                                     const std::string& token)
    : gateway_url_(gateway_url),
      token_(token),
      connection_retry_{.max_attempts = 3,
                        .base_delay_seconds = 0.1,
                        .max_delay_seconds = 1.0,
                        .jitter_ratio = 0.0},
      connection_breaker_(3, 5.0),
      simulation_mode_(gateway_url.find("localhost") != std::string::npos) {}

auto SwarmOrchestrator::execute_playbook(const nlohmann::json& playbook,
                                         const std::string& goal)
    -> std::vector<nlohmann::json> {
    spdlog::info("Starting swarm playbook: {}",
                 playbook.value("name", "unnamed"));

    for (const auto& phase : playbook.value("phases", nlohmann::json::array())) {
        execute_phase(phase, goal);
    }

    spdlog::info("Swarm playbook execution complete.");
    return history_;
}

void SwarmOrchestrator::execute_phase(const nlohmann::json& phase,
                                      const std::string& goal) {
    auto name = phase.value("name", "Unknown Phase");
    auto mode = phase.value("mode", "sequential");
    spdlog::info("Executing phase: {} (mode: {})", name, mode);

    std::vector<SwarmTask> tasks;
    for (const auto& d : phase.value("delegates", nlohmann::json::array())) {
        std::string content = d.value("task_template", "");
        auto pos = content.find("${goal}");
        if (pos != std::string::npos) {
            content.replace(pos, 7, goal);
        }
        tasks.push_back({
            .agent_id = d.value("agent", ""),
            .task_template = content,
            .result = {},
            .run_id = generate_run_id(),
        });
    }

    if (mode == "parallel") {
        execute_phase_parallel(tasks);
    } else {
        execute_phase_sequential(tasks);
    }
}

void SwarmOrchestrator::execute_phase_sequential(
    std::vector<SwarmTask>& tasks) {
    for (auto& task : tasks) {
        run_task(task);
    }
}

void SwarmOrchestrator::execute_phase_parallel(
    std::vector<SwarmTask>& tasks) {
    std::vector<std::future<void>> futures;
    futures.reserve(tasks.size());

    for (auto& task : tasks) {
        futures.push_back(std::async(std::launch::async, [this, &task] {
            run_task(task);
        }));
    }

    for (auto& f : futures) {
        f.get();
    }
}

void SwarmOrchestrator::run_task(SwarmTask& task) {
    auto provider = router_.select_provider(task.complexity);
    spdlog::info("Dispatching task to agent {} via {}...",
                 task.agent_id, provider);
    task.status = "running";

    if (simulation_mode_) {
        task.status = "completed";
        task.result = "Simulated result from " + task.agent_id;
        router_.track_usage(provider, 500);
        record_task(task, provider);
        return;
    }

    // Non-simulation: dispatch via WebSocket to gateway
    auto ws_url = derive_ws_url(gateway_url_);
    WebSocketClient client(ws_url);
    client.connect();

    if (!client.is_connected()) {
        task.status = "failed";
        task.result = "WebSocket connection failed to " + ws_url;
        record_task(task, provider);
        return;
    }

    nlohmann::json dispatch_msg = {
        {"type", "dispatch"},
        {"agent", task.agent_id},
        {"task", task.task_template},
        {"run_id", task.run_id},
        {"provider", provider},
    };

    auto response = client.send_and_wait(dispatch_msg);
    if (response && response->value("status", "") == "accepted") {
        task.status = "completed";
        task.result = response->dump();
        router_.track_usage(provider, 500);
    } else {
        task.status = "failed";
        task.result = response ? response->dump() : "No response from gateway";
    }
    record_task(task, provider);
}

void SwarmOrchestrator::record_task(const SwarmTask& task,
                                     const std::string& provider) {
    std::lock_guard lock(history_mutex_);
    history_.push_back({
        {"agent", task.agent_id},
        {"provider", provider},
        {"task", task.task_template},
        {"result", *task.result},
    });
}

} // namespace euxis::core
